### PR TITLE
initial coqdoc support

### PR DESCRIPTION
**Pull request by Alizter**

### CODE REVIEW

**Comment by rgrinberg** [comment]
@Alizter try running the tests again in this PR.

**Comment by rgrinberg** [comment]
Looks good. Now for the mundane:

- [ ] CHANGES entry
- [ ] description of the feature in the manual
- [ ] Fix CI

File: src/dune_rules/gen_rules.ml [hunk 1]
```diff
@@ -392,6 +410,10 @@ let gen_rules ~sctx ~dir components : Build_config.gen_rules_result Memo.t =
       >>= function
       | Group_part _ -> redirect_to_parent
       | Standalone_or_root { directory_targets; contents } ->
+        let directory_targets =
+          (* [dir] isn't quite right *)
+          collect_directory_targets sctx ~dir ~init:directory_targets
```
Comment (Alizter) [inline src/dune_rules/gen_rules.ml:415]: @rgrinberg I don't remember, did we decide this was fine in the end?

**Comment by Alizter** [comment]
Little tricky to know what is right. Compare:

https://github.com/coq/coq/blob/6eda65e288831a27ef2edb7607ed2d01dffb3975/tools/CoqMakefile.in#L543-L547

with

https://github.com/coq/coq/blob/6eda65e288831a27ef2edb7607ed2d01dffb3975/tools/CoqMakefile.in#L829-L831

The coq_makefile approach seems to be both incrementally making HTML files and globbing them all into one operation. The latter is done for a table-of-contents --toc. Currently in dune we are doing only the latter. I don't see a way to make this incremental tbh.

**Comment by ejgallego** [comment]
@Alizter I guess we'll have to look at the code of Coqdoc to figure out how it works, not surprising :smiling_imp:

**Comment by ejgallego** [comment]
Incidentally my first PR to Coq was on coqdoc, sadly it wasn't merged https://github.com/coq/coq/pull/133

**Comment by Alizter** [comment]
@ejgallego Incremental builds would be nice to have but I will leave it for a future PR. This is ready to merge.

**Comment by rgrinberg** [comment]
Unfortunately this PR requires the fix in https://github.com/ocaml/dune/pull/5754 which is not yet done correctly.

**Comment by rgrinberg** [comment]
@Alizter could you squash into a single commit?

**Comment by Alizter** [comment]
FTR doc changes will come later in https://github.com/ocaml/dune/pull/5777

File: src/dune_rules/gen_rules.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_rules/gen_rules.ml:415]: yes, it's fine.

Comment (Alizter) [reply src/dune_rules/gen_rules.ml:415]: Shall I clean up that comment then?

Comment (rgrinberg) [reply src/dune_rules/gen_rules.ml:415]: Yes

File: CHANGES.md [hunk 2]
```diff
@@ -9,6 +9,11 @@ Unreleased
 
 - Support new locations of unix, str, dynlink in OCaml >= 5.0 (#5582, @dra27)
 
+- The ``coq.theory`` stanza now produces rules for running ``coqdoc``. Given a
+  theory named ``mytheory``, the directory targets ``mytheory.html/`` and
+  ``mytheory.tex/`` or additionally the aliases `@doc` and `@doc-latex` will
+  build the HTML and LaTeX documentation repsectively. (#5695, @Alizter)
```
Comment (rgrinberg) [inline CHANGES.md:15]: also add `fixes #3760` here.

**Pull request merged**